### ISSUE TITLE
Allow `expression` in code template to be conditional

### ISSUE THREAD

**Issue by Piinks** [dart-lang/sdk#47943]
Dart fix allows for code templates, and within a given code template expression, variables can be inserted to generate new code for the user.
This is super powerful but becomes quickly increases in complexity when the expression you want to choose is based on multiple variables of the old code.

For example, from the [the button universe migration ](https://docs.flutter.dev/release/breaking-changes/buttons), the `FlatButton class has 14 different parameters that could be set that should be wrapped up into a new ButtonStyle in the new TextButton.
In order to migrate all of these, the fix rule (AFAIK) has to enumerate every combination of these parameters being set or not to properly migrate the code. This would be 16,383 rules.

It would be nice if in addition to inserting variables, the expression could also support conditionals so if a parameter is not set, we can just not add that line in the code template expression. It may also be possible to do this another way, I just may not be aware of it.

## Example

We already have an example of this in `fix_data.yaml`, in this case it was 3 parameters so it was not too bad to enumerate all of the combinations. But the case with button and 14 parameters would be too many.

<details> 

  <summary>Current code enumerating parameter combinations </summary>

```yaml
# Changes made in https://github.com/flutter/flutter/pull/66482
  - title: "Migrate to 'TextSelectionThemeData'"
    date: 2020-09-24
    element:
      uris: [ 'material.dart' ]
      constructor: 'raw'
      inClass: 'ThemeData'
    oneOf:
      - if: "textSelectionColor != '' && cursorColor != '' && textSelectionHandleColor != ''"
        changes:
          - kind: 'addParameter'
            index: 73
            name: 'textSelectionTheme'
            style: optional_named
            argumentValue:
              expression: 'TextSelectionThemeData(cursorColor: {% cursorColor %}, selectionColor: {% textSelectionColor %}, selectionHandleColor: {% textSelectionHandleColor %},)'
              requiredIf: "textSelectionColor != '' && cursorColor != '' && textSelectionHandleColor != ''"
          - kind: 'removeParameter'
            name: 'textSelectionColor'
          - kind: 'removeParameter'
            name: 'cursorColor'
          - kind: 'removeParameter'
            name: 'textSelectionHandleColor'
          - kind: 'removeParameter'
            name: 'useTextSelectionTheme'
      - if: "textSelectionColor == '' && cursorColor != '' && textSelectionHandleColor != ''"
        changes:
          - kind: 'addParameter'
            index: 73
            name: 'textSelectionTheme'
            style: optional_named
            argumentValue:
              expression: 'TextSelectionThemeData(cursorColor: {% cursorColor %}, selectionHandleColor: {% textSelectionHandleColor %},)'
              requiredIf: "textSelectionColor == '' && cursorColor != '' && textSelectionHandleColor != ''"
          - kind: 'removeParameter'
            name: 'cursorColor'
          - kind: 'removeParameter'
            name: 'textSelectionHandleColor'
          - kind: 'removeParameter'
            name: 'useTextSelectionTheme'
      - if: "textSelectionColor != '' && cursorColor != '' && textSelectionHandleColor == ''"
        changes:
          - kind: 'addParameter'
            index: 73
            name: 'textSelectionTheme'
            style: optional_named
            argumentValue:
              expression: 'TextSelectionThemeData(cursorColor: {% cursorColor %}, selectionColor: {% textSelectionColor %},)'
              requiredIf: "textSelectionColor != '' && cursorColor != '' && textSelectionHandleColor == ''"
          - kind: 'removeParameter'
            name: 'textSelectionColor'
          - kind: 'removeParameter'
            name: 'cursorColor'
          - kind: 'removeParameter'
            name: 'useTextSelectionTheme'
      - if: "textSelectionColor != '' && cursorColor == '' && textSelectionHandleColor != ''"
        changes:
          - kind: 'addParameter'
            index: 73
            name: 'textSelectionTheme'
            style: optional_named
            argumentValue:
              expression: 'TextSelectionThemeData(selectionColor: {% textSelectionColor %}, selectionHandleColor: {% textSelectionHandleColor %},)'
              requiredIf: "textSelectionColor != '' && cursorColor == '' && textSelectionHandleColor != ''"
          - kind: 'removeParameter'
            name: 'textSelectionColor'
          - kind: 'removeParameter'
            name: 'textSelectionHandleColor'
          - kind: 'removeParameter'
            name: 'useTextSelectionTheme'
      - if: "textSelectionColor == '' && cursorColor != '' && textSelectionHandleColor == ''"
        changes:
          - kind: 'addParameter'
            index: 73
            name: 'textSelectionTheme'
            style: optional_named
            argumentValue:
              expression: 'TextSelectionThemeData(cursorColor: {% cursorColor %})'
              requiredIf: "textSelectionColor == '' && cursorColor != '' && textSelectionHandleColor == ''"
          - kind: 'removeParameter'
            name: 'cursorColor'
          - kind: 'removeParameter'
            name: 'useTextSelectionTheme'
      - if: "textSelectionColor != '' && cursorColor == '' && textSelectionHandleColor == ''"
        changes:
          - kind: 'addParameter'
            index: 73
            name: 'textSelectionTheme'
            style: optional_named
            argumentValue:
              expression: 'TextSelectionThemeData(selectionColor: {% textSelectionColor %})'
              requiredIf: "textSelectionColor != '' && cursorColor == '' && textSelectionHandleColor == ''"
          - kind: 'removeParameter'
            name: 'textSelectionColor'
          - kind: 'removeParameter'
            name: 'useTextSelectionTheme'
      - if: "textSelectionColor == '' && cursorColor == '' && textSelectionHandleColor != ''"
        changes:
          - kind: 'addParameter'
            index: 73
            name: 'textSelectionTheme'
            style: optional_named
            argumentValue:
              expression: 'TextSelectionThemeData(selectionHandleColor: {% textSelectionHandleColor %})'
              requiredIf: "textSelectionColor == '' && cursorColor == '' && textSelectionHandleColor != ''"
          - kind: 'removeParameter'
            name: 'textSelectionHandleColor'
          - kind: 'removeParameter'
            name: 'useTextSelectionTheme'
      - if: "useTextSelectionTheme != ''"
        changes:
          - kind: 'removeParameter'
            name: 'useTextSelectionTheme'
    variables:
      textSelectionColor:
        kind: 'fragment'
        value: 'arguments[textSelectionColor]'
      cursorColor:
        kind: 'fragment'
        value: 'arguments[cursorColor]'
      textSelectionHandleColor:
        kind: 'fragment'
        value: 'arguments[textSelectionHandleColor]'
      useTextSelectionTheme:
        kind: 'fragment'
        value: 'arguments[useTextSelectionTheme]'
```

</details>

<details>
  
  <summary>Proposed functionality</summary>

Instead of going through every combination, can we add options to the expression?
Like if any of these are set, use this expression, and it decides what to apply or not.

```yaml
# Changes made in https://github.com/flutter/flutter/pull/66482
  - title: "Migrate to 'TextSelectionThemeData'"
    date: 2020-09-24
    element:
      uris: [ 'material.dart' ]
      constructor: 'raw'
      inClass: 'ThemeData'
    oneOf:
      # If any of these are set, apply in conditional expression
      - if: "textSelectionColor != '' || cursorColor != '' || textSelectionHandleColor != ''"
        changes:
          - kind: 'addParameter'
            index: 73
            name: 'textSelectionTheme'
            style: optional_named
            argumentValue:
              # Syntax is probably way off here, but something like this,
              # if this value wasn't set in the old code, skip it.
              expression: 'TextSelectionThemeData('
                "cursorColor != ''" ? 'cursorColor: {% cursorColor %},' : ''
                "selectionColor != ''" ? 'selectionColor: {% textSelectionColor %},' : ''
                "selectionHandleColor != ''" ? 'selectionHandleColor: {% textSelectionHandleColor %},' : ''
              ')'
              requiredIf: "textSelectionColor != '' || cursorColor != '' || textSelectionHandleColor != ''"
          - kind: 'removeParameter'
            name: 'textSelectionColor'
          - kind: 'removeParameter'
            name: 'cursorColor'
          - kind: 'removeParameter'
            name: 'textSelectionHandleColor'
          - kind: 'removeParameter'
            name: 'useTextSelectionTheme'
    variables:
      textSelectionColor:
        kind: 'fragment'
        value: 'arguments[textSelectionColor]'
      cursorColor:
        kind: 'fragment'
        value: 'arguments[cursorColor]'
      textSelectionHandleColor:
        kind: 'fragment'
        value: 'arguments[textSelectionHandleColor]'
```

</details>

**Comment by Piinks** [comment]
cc @bwilkerson

**Comment by Piinks** [comment]
These are all pretty closely related, linking for reference:
- https://github.com/flutter/flutter/issues/95284
- #47945
- #47944
- #47941

**Comment by bwilkerson** [comment]
The combinatorial explosion is definitely a problem we need to solve, and this is an interesting approach. I'm not sure what kind of syntax we'd be able to have because I'm not that familiar with all of the details of YAML, but we ought to be able to do something along those lines.

If we can't have the conditional expressions in-lined (which I'd prefer) we ought to be able to do something similar. For example, we might define a new kind of variable:
```yaml
variables:
  cursorColorArgument:
    kind: 'conditional'
    oneOf:
      - if: cursorColor != ''
        value: 'cursorColor: {% cursorColor %},'
      - else:
        value: ''
```

**Comment by bwilkerson** [comment]
> Well, why not use Dart

Two reasons that I can think of at the moment; there might be more.

First, it would defeat one of the basic design principles we used for data-driven fixes, which is that users shouldn't need to understand the corners of the language in order to produce valid results; that should be on us as implementors of the framework. Even for something as simple as the code you wrote above, you're missing a check to see whether we need to have a trailing comma at the end of the argument list. That shouldn't be something that the Flutter framework authors have to think about, let alone know how to do, in order to help their users migrate forward.

Second, we can't dynamically load arbitrary Dart code, which means we'd have to write an interpreter that could read the Dart code and emulate what the Dart VM would have compiled the code to do. That's a very big task that we don't want to take on, or even accept the burden of maintaining if someone else were to take it on.

**Comment by asashour** [comment]
Thanks for the feedback.

My sample code was possibly pointing to a different direction, so please let me rephrase.

There are two things, the condition defined by the user in order to apply a fix, and the implementation details of that fix (which should be handled by the `data-driven` code in the analyzer). I didn't mean to talk about second point.

As I understood from this issue, is that the condition where the current offer is not powerful enough, and the user needs to put the value to be a "static" strings controlled by user conditions, those for example:

```dart
expression: 'TextSelectionThemeData('
                "cursorColor != ''" ? 'cursorColor: {% cursorColor %},' : ''
                "selectionColor != ''" ? 'selectionColor: {% textSelectionColor %},' : ''
                "selectionHandleColor != ''" ? 'selectionHandleColor: {% textSelectionHandleColor %},' : ''
              ')'
```


The next part is based on the assumption that the user is writing the output code `expression`, and this is not something done by `data-drive` part, because otherwise we need more possibilities of adding expressions.


There seems to be [evaluateIn](https://github.com/dart-lang/sdk/blob/cfddeee68e0cf7451a2eee1a9fbbff8a8d9c0b92/pkg/analysis_server/lib/src/services/correction/fix/data_driven/modify_parameters.dart#L84), which seems to evaluate arbitrary code string, and it handles very few cases, but I guess it can be enhanced to support cases like:


```dart
expression: |
    String f(String? cursorColor, String? selectionColor) {
      return 'TextSelectionThemeData() '
        '${cursorColor != null ? 'cursorColor: $cursorColor,' : ''}'
        '${selectionColor != null ? 'selectionColor : $selectionColor,' : ''}';
  }
```

or 

```dart
String f2(String? cursorColor, String? selectionColor) {
  var s = 'TextSelectionThemeData(';
  if (cursorColor != null) {
    s += 'cursorColor: $cursorColor,';
  }
  if (selectionColor != null) {
    s += 'selectionColor: selectionColor,';
  }
  return '$s);';
}
```


Another possibility is to split defining "variables" and use those variables in the string, something like:

```dart
  var1:
    cursorColor != '' ? 'cursorColor: $cursorColor' : ''
  var2:
    selectionColor != '' ? 'selectionColor: $selectionColor' : ''
```

And using it in the expression of

```
  expression: |
    TextSelectionThemeData($cursorColor$selectionColor)
```

Using dart in the condition could not be harmful, since the user is a library author, and no need to parse and document other ways to define complex conditions.

Another idea if we don't want to do any parsing, is to write code to a temporary file, execute it by `dart` and get the output of the execution.

**Comment by bwilkerson** [comment]
> The next part is based on the assumption that the user is writing the output code `expression` ...

That is largely valid. The exception is when the output depends on factors that the fix author shouldn't need to worry about, such as adding required imports, or using the correct syntax based on language changes. That's why we have variable values of kind 'import'.

> There seems to be `evaluateIn`, which seems to evaluate arbitrary code string, and it handles very few cases, but I guess it can be enhanced to support cases like:

It doesn't handle arbitrary code strings; it handles a very small expression syntax that is consistent with Dart but is very limited (see https://github.com/dart-lang/sdk/blob/main/pkg/analysis_server/lib/src/services/correction/fix/data_driven/expression.dart). Yes, we can enhance it to support more cases. That's what I meant by "have the conditional expressions in-lined". If YAML doesn't already define a meaning to `?:` that would conflict with the syntax you suggested (`'${cursorColor != null ? 'cursorColor: $cursorColor,' : ''}'`) then we'd be able to support that (with the exception that the outer quotes would need to be double quotes). And as I said, that would be my preference.

> or by having a complete function, which can be copied and pasted from the user IDE.

That's what I meant by my second point:

> Second, we can't dynamically load arbitrary Dart code, which means we'd have to write an interpreter that could read the Dart code and emulate what the Dart VM would have compiled the code to do.

Supporting complete functions would require writing a compiler and interpreter for the whole Dart language, and that isn't feasible.

> Another possibility is to split defining "variables" and use those variables in the string ...

I believe that doing that would require the same amount of work as simply supporting the expressions in the templates (and we might even get support in both places automatically, I'd have to look more closely at the code), and I think that allowing it in the templates would be better for fix authors (assuming we have to choose between the two options).

> Another idea if we don't want to do any parsing, is to write code to a temporary file, execute it by dart and get the output of the execution.

Unfortunately, we can't afford the performance hit of starting a separate process (or isolate) to execute the code.

> At the end, I guess it should be decided how the user would best interact with the fix, and the implementation follows.

Absolutely! Baring technical problems, I think that extending the expression syntax to handle conditional expressions might well be the best solution from a fix author's perspective. But one purpose of this thread is to hear feedback from those authors to see what they would prefer.

@Piinks In case you have an opinion here.
